### PR TITLE
YaRN support implementation

### LINKED ISSUE #980: Support YaRN models (RoFormer implementation in rotary_embedding kernel)
The YaRN model with a context size of 64k and 128k was recently released and pre-trained by people from Nous Research and EleutherAI. It uses the RoFormer type of embeddings that seem different from GPT-NeoX and GPT-J. It is based on the LLaMa 2 model, so support is mostly there, just need some small adjustments.

The original [YaRN module](https://huggingface.co/conceptofmind/Yarn-Llama-2-13b-128k/blob/main/modeling_llama_together_yarn.py#L116) uses the [flash attention rotary embedding](https://github.com/Dao-AILab/flash-attention/blob/a1576ad1e887c11f4b76f42e9dfaceeb6369cdb8/csrc/rotary/rotary_cuda.cu#L9) implementation and seems similar in functionality. You may also be interested in the original RoFormer implementation from [Huggingface](https://github.com/huggingface/transformers/blob/df04959e5542d41b269f96305d82c62287350cee/src/transformers/models/roformer/modeling_roformer.py#L319).

Models catalog:
https://huggingface.co/NousResearch/Yarn-Llama-2-7b-64k
https://huggingface.co/NousResearch/Yarn-Llama-2-7b-128k
https://huggingface.co/NousResearch/Yarn-Llama-2-13b-64k
https://huggingface.co/NousResearch/Yarn-Llama-2-13b-128k

### PR DESCRIPTION
**Pull request by Yard1**
Supersedes https://github.com/vllm-project/vllm/pull/1161, thank you @viktor-ferenczi for laying down the groundwork :)

This PR implements support for YaRN models.

YaRN paper: https://arxiv.org/abs/2309.00071
YaRN repository: https://github.com/jquesnelle/yarn
Smallest model to test with: https://huggingface.co/NousResearch/Yarn-Llama-2-7b-64k

Closes https://github.com/vllm-project/vllm/issues/980

### CODE REVIEW

**Comment by Yard1** [comment]
cc @WoosukKwon @zhuohan123

**Comment by viktor-ferenczi** [comment]
Use `format.sh` to format the code (it is in the repository in the top folder).

**Comment by viktor-ferenczi** [comment]
Tested it, but got the following error with model [NousResearch/Yarn-Llama-2-7b-64k](https://huggingface.co/NousResearch/Yarn-Llama-2-7b-64k/):
```
ValueError: `rope_scaling` must be a dictionary with with two fields, `type` and `factor`, got {'factor': 16.0, 'original_max_position_embeddings': 4096, 'type': 'yarn', 'finetuned': True}
```

**Comment by Yard1** [comment]
@viktor-ferenczi can you post the full stack trace?

**Comment by Yard1** [comment]
@viktor-ferenczi were you loading from a directory or from HF model id? I think it may not work in the former case as HF Transformers will use the built-in LlamaConfig class, which causes validation to fail. It should work if you specify the HF model id instead, as that should use the configuration present in the repository. This is a HF Transformers issue.

**Comment by viktor-ferenczi** [comment]
Thanks for the diagnosis. 

I tried it with both a MODEL_DIR and a MODEL_ID, both fails with the same exception:
```
/home/viktor/dep/vllm-contrib/venv/bin/python /home/viktor/dep/vllm-contrib/tests/yarn/model_llama2_7b_vllm.py 
Downloading (…)lve/main/config.json: 100%|██████████| 899/899 [00:00<00:00, 6.09MB/s]
Traceback (most recent call last):
  File "/home/viktor/dep/vllm-contrib/tests/yarn/model_llama2_7b_vllm.py", line 53, in <module>
    main()
  File "/home/viktor/dep/vllm-contrib/tests/yarn/model_llama2_7b_vllm.py", line 41, in main
    model = Model()
  File "/home/viktor/dep/vllm-contrib/tests/yarn/model_llama2_7b_vllm.py", line 17, in __init__
    self.llm = LLM(
  File "/home/viktor/dep/vllm-contrib/vllm/entrypoints/llm.py", line 93, in __init__
    self.llm_engine = LLMEngine.from_engine_args(engine_args)
  File "/home/viktor/dep/vllm-contrib/vllm/engine/llm_engine.py", line 225, in from_engine_args
    engine_configs = engine_args.create_engine_configs()
  File "/home/viktor/dep/vllm-contrib/vllm/engine/arg_utils.py", line 182, in create_engine_configs
    model_config = ModelConfig(self.model, self.tokenizer,
  File "/home/viktor/dep/vllm-contrib/vllm/config.py", line 79, in __init__
    self.hf_config = get_config(model, trust_remote_code, revision)
  File "/home/viktor/dep/vllm-contrib/vllm/transformers_utils/config.py", line 42, in get_config
    raise e
  File "/home/viktor/dep/vllm-contrib/vllm/transformers_utils/config.py", line 30, in get_config
    config = AutoConfig.from_pretrained(
  File "/home/viktor/dep/vllm-contrib/venv/lib/python3.10/site-packages/transformers/models/auto/configuration_auto.py", line 1051, in from_pretrained
    return config_class.from_dict(config_dict, **unused_kwargs)
  File "/home/viktor/dep/vllm-contrib/venv/lib/python3.10/site-packages/transformers/configuration_utils.py", line 747, in from_dict
    config = cls(**config_dict)
  File "/home/viktor/dep/vllm-contrib/venv/lib/python3.10/site-packages/transformers/models/llama/configuration_llama.py", line 152, in __init__
    self._rope_scaling_validation()
  File "/home/viktor/dep/vllm-contrib/venv/lib/python3.10/site-packages/transformers/models/llama/configuration_llama.py", line 171, in _rope_scaling_validation
    raise ValueError(
ValueError: `rope_scaling` must be a dictionary with with two fields, `type` and `factor`, got {'factor': 16.0, 'original_max_position_embeddings': 4096, 'type': 'yarn', 'finetuned': True}
```

I've just removed the `len(self.rope_scaling) != 2` check temporarily from `transformers/models/llama/configuration_llama.py`:
```python
if not isinstance(self.rope_scaling, dict) or len(self.rope_scaling) != 2:
```

File: tests/yarn/model_llama2_7b_vllm.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+    def __init__(self):
+        super().__init__()
+        self.model_id = MODEL_ID
+        self.llm = LLM(
+            model=MODEL_ID,
+            max_model_len=8192,
+            max_num_batched_tokens=8192,
+            seed=42,
+        )
```
Comment (viktor-ferenczi) [inline tests/yarn/model_llama2_7b_vllm.py:22]: Add this parameter:
`trust_remote_code=True,`

It is required to bypass the checks in the transformers library.

**Comment by viktor-ferenczi** [comment]
Tested OK with the `NousResearch/Yarn-Llama-2-7b-64k` model up to 8192 context length using the `pass_key_evaluator.py` test script.

**Comment by viktor-ferenczi** [comment]
Attempted to test at full 64k context length:
```
            max_model_len=65536,
            max_num_batched_tokens=65536,
```

Ran into this error:
```
  File "/home/viktor/dep/vllm-contrib/vllm/worker/worker.py", line 381, in _check_if_can_support_max_seq_len
    raise RuntimeError(
RuntimeError: vLLM cannot currently support max_model_len=65536 with block_size=16 on GPU with compute capability (8, 9) (required shared memory 264252.0 > available shared memory 101376). This will be fixed in a future release.
```

**Comment by viktor-ferenczi** [comment]
More test results:
- 8448: Runs, but cannot find the pass key
- 12288: OK
- 13312: OK
- 13824: Runs, but cannot find the pass key
- 14440: Frozen
- 16384: Frozen
- 25328: Frozen
- Above 25328: Error (see above)

(4090 GPU with 24GB VRAM)

File: tests/yarn/pass_key_evaluator.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+            "".join(fragments),
+            pass_key,
+            self.model.count_tokens(garbage_prefix),
+        )
+
+    def evaluate(self,
+                 max_tokens: int,
+                 resolution: int = 100,
+                 n: int = 10) -> Iterable[Tuple[int, int, int]]:
+        assert max_tokens > 0
```
Comment (viktor-ferenczi) [inline tests/yarn/pass_key_evaluator.py:57]: Add parameter: `headroom: int = 8`

File: tests/yarn/pass_key_evaluator.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        assert resolution > 1
+
+        garbage_count = max_tokens // self.model.count_tokens(self.garbage)
```
Comment (viktor-ferenczi) [inline tests/yarn/pass_key_evaluator.py:60]: Reserve headroom, so the generated key can fit into the model's context window in all cases:
```py
garbage_count = (max_tokens - headroom) // self.model.count_tokens(self.garbage)
```

Comment (viktor-ferenczi) [inline tests/yarn/pass_key_evaluator.py:60]: Reserve headroom:
```python
        garbage_count = (max_tokens - headroom) // self.model.count_tokens(self.garbage)
        while (garbage_count and self.model.count_tokens(
                self.format_prompt(garbage_count, 0)[0]) > max_tokens - headroom):
            garbage_count -= 1
        assert garbage_count
```

**Comment by Yard1** [comment]
I think it might be frozen because you do not have enough KV cache memory to support the entire context length. This is orthogonal to the error raised about shared memory. The only "fix" here would be to use a GPU with more memory (and perhaps raise a warning in the scheduler if a new request cannot be scheduled due to lack of KV cache).

**Comment by viktor-ferenczi** [comment]
I'm going to try on both 4090s then, I have two :)

Another way is to increase `swap_space` from the default 4 (GB) to higher, which adds more CPU memory to the cache.

**Comment by viktor-ferenczi** [comment]
Tested OK on two 4090 GPUs up to 25088 context size. (Not exhaustively, just at various points.)

**Comment by viktor-ferenczi** [comment]
@casper-hansen We got progress with YaRN here, but facing problems/limitations with vLLM at longer context sizes. Could you please suggest? Thanks!

**Comment by Yard1** [comment]
I think memory related context length limitations should not block this PR and can be fixed independently. I know that the vLLM team is investigating ways to improve shared memory usage, for example.

**Comment by viktor-ferenczi** [comment]
I agree about not blocking this PR because of unrelated context length limitations.

I'm not sure how #555 was tested, cannot find its test case in the source code. Wanted to reuse that test case for testing this PR as well, since they are related. (I could test manually with the pass key retrieval test I added, but it is not streamlined, not added to automation.)

**Comment by casper-hansen** [comment]
> @casper-hansen We got progress with YaRN here, but facing problems/limitations with vLLM at longer context sizes. Could you please suggest? Thanks! - See [#1264 (comment)](https://github.com/vllm-project/vllm/pull/1264#issuecomment-1751802168)

I would advise to test on A100s to make sure it’s not a limit on the memory side. Other than that, it’s probably some niche area of vLLM that I have little insight into.

**Comment by viktor-ferenczi** [comment]
#555 was similar, but I cannot find the test case to cover the long context window there. It would be nice to reuse it. As far as I remember they had a different variant of pass key retrieval test. We should also add the test to automation, so it is running on the pipeline.

**Comment by WoosukKwon** [comment]
@Yard1 Would it be possible to first add Yarn without the tests? We also need to add tests for other RoPe scaling methods and it seems the Yarn test code will conflict with that.

**Comment by Yard1** [comment]
@WoosukKwon sure, I can separate the tests out into another PR

**Comment by Yard1** [comment]
@WoosukKwon done

File: tests/__init__.py
Comment (WoosukKwon) [inline tests/__init__.py:1]: A dumb question: What is the purpose of having this file?

Comment (Yard1) [reply tests/__init__.py:1]: ah could be a left over, let me remove it

**Comment by WoosukKwon** [comment]
@Yard1 I got this error when tried the `NousResearch/Yarn-Llama-2-7b-64k` model. Is this because I'm using transformers 4.34.0?

```
  File "/home/wskwon/anaconda3/envs/vllm/lib/python3.9/site-packages/transformers/models/llama/configuration_llama.py", line 171, in _rope_scaling_validation
    raise ValueError(
ValueError: `rope_scaling` must be a dictionary with with two fields, `type` and `factor`, got {'factor': 16.0, 'original_max_position_embeddings': 4096, 'type': 'yarn', 'finetuned': True}
```

**Comment by Yard1** [comment]
@WoosukKwon did you use `trust_remote_code=True`? It is required here.

**Comment by WoosukKwon** [comment]
@Yard1 Got it. It works after adding `trust_remote_code`.

However, I got the following error during the initial memory profiling:
```
  File "/home/wskwon/workspace/vllm/vllm/model_executor/layers/attention.py", line 203, in forward
    output = torch.empty_like(query)
RuntimeError: CUDA error: an illegal memory access was encountered
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions
```
However, the error disappeared when I fixed `max_model_len` to a smaller number (e.g., 4096) than the default one (1M). Have you seen this too?

We saw the same kind of error when using Mistral model with AWQ (#1236). The error happened because there was integer overflow in pointer arithmetic  inside the AWQ GEMM kernel, when the number of input tokens is too large. We fixed this in #1295 by changing some `int` variables to `long long`. Likewise, I feel one of our kernels does not work with 1M input tokens and silently reads/writes invalid memory address.

**Comment by Yard1** [comment]
Yes, I think the default transformers input len is way too big. I got it to work successfully with 24K on an A100.

**Review by viktor-ferenczi — APPROVED** [review]

File: vllm/model_executor/layers/attention.py [hunk 4]
```diff
@@ -334,6 +334,19 @@ def __init__(
                 self.rotary_emb = DynamicNTKScalingRotaryEmbedding(
                     head_size, rotary_dim, max_position, base, is_neox_style,
                     scaling_factor)
+            elif scaling_type == "yarn":
+                new_max_position = rope_scaling[
```
Comment (WoosukKwon) [inline vllm/model_executor/layers/attention.py:338]: Is the variable name `new_max_position`? If I understand correctly, this is the original maximum length (e.g., 4096 for `NousResearch/Yarn-Llama-2-7b-64k`).

**Review by WoosukKwon — COMMENTED** [review]
It seems the configs of Yarn models break https://github.com/vllm-project/vllm/blob/9738b84a08957eb828669e8af27337ee722e8fdc/vllm/config.py#L354

For example, vLLM infers that the maximum length of `NousResearch/Yarn-Llama-2-7b-64k` is 1M rather than 64K because the `max_position_embeddings` in the model config is 64K and the rope scaling factor is 16.

On the other hand, for `NousResearch/Yarn-Mistral-7b-128k` model, vLLM regards the maximum length is 512K because `max_position_embeddings` in its config is 32K (which is the original model's maximum length after 4x RoPE scaling).

In conclusion, I feel we need a separate logic for Yarn models in the `_get_and_verify_max_len` function. WDYT?

File: vllm/model_executor/layers/attention.py (hunk 4 above)
Comment (Yard1) [reply vllm/model_executor/layers/attention.py:338]: Let me take a look at both comments!

**Comment by Yard1** [comment]
@WoosukKwon Updated, PTAL!

**Comment by Yard1** [comment]
I discovered some sort of a correctness error with the probability tensor getting a non-finite value, marking as draft until investigation is complete.

**Comment by Yard1** [comment]
Fixed, the int in kernel was overflowing.

**Comment by casper-hansen** [comment]
Hi @Yard1, just informing you that there is now a Mistral Yarn model out. Perhaps a good idea to test if this PR works with it.
https://huggingface.co/NousResearch/Yarn-Mistral-7b-64k

**Comment by Yard1** [comment]
Good catch with the activation kernel @WoosukKwon !

**Review by WoosukKwon — APPROVED** [review]
@Yard1 LGTM. Thanks for submitting the PR! FYI, I also fixed the same kind of overflow error in `activation_kernels.cu`.

**Comment by xiechengmude** [comment]
BTw whats the minimal resourses to deploy this 64k model ?
